docs/common/craft-parts/explanation/migration.rst

```diff
@@ -0,0 +1,115 @@
+from a shared directory, or when files from a specific part must be removed
+from an area containing files from different parts. This mapping is obtained
+when migrating files from part-specific directories, and is called the
+"migration state".
+
+
+Simple file migration
+---------------------
+
+During the execution of a simple four-step lifecycle, files from the part's
+install directory are migrated to the stage area and from there to prime.
+In the following diagram, files are represented as circles with letters. The
+green circles correspond to files from part 1, while blue circles represent
+files originating from part 2.
+
+.. image:: /common/craft-parts/images/simple_migration.png
+
+Files ``A`` and ``B`` originate from part 1, while files ``C`` and ``D``
+originate from part 2. They are staged into a common area, and then migrated
+to prime. Note that file ``D`` is filtered out when priming, and is not part
+of the final primed contents. The migration state for parts 1 and 2 is obtained
+from the contents of each of the parts' install directories while they're still
+separate (marked with red ellipses in the diagram).
+
+When files from a specific part need to be migrated from stage to prime or
+removed from stage or prime, the migration state contains the list of files
+to migrate or remove.
+
+
+File migration using partitions
+-------------------------------
+
+When the partitions feature is enabled, files can be organized from a part's
+install directory to a different partition. In this case, files are migrated
+to the partition's specific stage and prime directories.
+
+.. image:: /common/craft-parts/images/partition_migration.png
+
+When partitions are used, an operation to remove or migrate files from a
+specific part happens across all partitions. In this example, if files from
+part 1 are to be removed from prime, files ``A`` and ``B`` will be deleted.
+For that, separate migration states are obtained from all parts in all
+partitions.
+
+
+File migration using overlays
+-----------------------------
+
+When the overlay feature is enabled, files can originate from both the part's
+install directory and the overlay area. In the diagram, files from the overlay
+are represented as squares, and files from the build install are represented
+as circles.
+
+.. image:: /common/craft-parts/images/overlay_migration.png
+
+If overlays are used, it's important to note that the files originating from
+the overlay form a set that are moved or deleted as a group. In the example
+above, if files from part 1 are to be deleted from prime, only files ``A``
+and ``B`` are removed. Overlay files are only removed when files from all parts
+containing overlay files are removed. Conversely, when files are migrated, the
+entire set of overlay files is migrated along with the first part that contains
+overlay files. A part with no overlay files will have all its files processed
+immediately.
+
+
+File migration using partitions and overlays
```

Comment:
This is the general case we'll need to cover in the lifecycle refactoring. All other scenarios should be particular cases where a type of migration won't happen.